File: android/app/src/main/java/com/collectrndemo/simple/modules/show/VGSShowModule.java
package com.collectrndemo.simple.modules.show;

import androidx.annotation.NonNull;

import com.collectrndemo.simple.modules.SharedConfig;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.verygoodsecurity.vgsshow.VGSShow;
import com.verygoodsecurity.vgsshow.core.VGSEnvironment;
import com.verygoodsecurity.vgsshow.core.logs.VGSShowLogger;
import com.verygoodsecurity.vgsshow.core.network.client.VGSHttpMethod;
import com.verygoodsecurity.vgsshow.core.network.model.VGSResponse;
import com.verygoodsecurity.vgsshow.widget.VGSTextView;

public class VGSShowModule extends ReactContextBaseJavaModule {

    private static ReactApplicationContext reactContext;

    private VGSShow show;

    private Callback callback;

    VGSShowModule(ReactApplicationContext c) {
        super(c);
        reactContext = c;
    }

    @Override
    public void initialize() {
        super.initialize();
        VGSShowLogger.INSTANCE.setLevel(VGSShowLogger.Level.DEBUG);
        show = new VGSShow.Builder(reactContext, SharedConfig.VAULT_ID)
                .setEnvironment(VGSEnvironment.Companion.toVGSEnvironment(SharedConfig.ENVIRONMENT))
                .build();
        show.addOnResponseListener(this::sendResponse);
    }

    @NonNull
    @Override
    public String getName() {
        return "VGSShowManager";
    }

    public void subscribe(VGSTextView view) {
        show.subscribe(view);
    }

    @ReactMethod
    public void revealData(Callback callback) {
        this.callback = callback;
        show.requestAsync("/post", VGSHttpMethod.POST, SharedConfig.showPayload);
    }

    private void sendResponse(VGSResponse response) {
        StringBuilder responseStr = new StringBuilder("Code: ")
                .append(response.getCode());
        if (response instanceof VGSResponse.Error) {
            responseStr
                    .append("\n")
                    .append("Error: ")
                    .append(((VGSResponse.Error) response).getMessage());
        }
        callback.invoke(responseStr.toString());
    }
}

File: android/app/src/main/kotlin/com/collectrndemo/simple/modules/show/VGSShowModule.kt
@file:Suppress("unused")

package com.collectrndemo.simple.modules.show

import com.collectrndemo.simple.modules.SharedConfig
import com.collectrndemo.simple.modules.SharedConfig.showPayload
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.verygoodsecurity.vgsshow.VGSShow
import com.verygoodsecurity.vgsshow.core.VGSEnvironment.Companion.toVGSEnvironment
import com.verygoodsecurity.vgsshow.core.listener.VGSOnResponseListener
import com.verygoodsecurity.vgsshow.core.logs.VGSShowLogger
import com.verygoodsecurity.vgsshow.core.logs.VGSShowLogger.level
import com.verygoodsecurity.vgsshow.core.network.client.VGSHttpMethod
import com.verygoodsecurity.vgsshow.core.network.model.VGSResponse
import com.verygoodsecurity.vgsshow.widget.VGSTextView

/** @noinspection unused
 */
class VGSShowModule internal constructor(
    reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext), VGSOnResponseListener {

    private lateinit var show: VGSShow

    private var callback: Callback? = null

    override fun initialize() {
        super.initialize()
        level = VGSShowLogger.Level.DEBUG
        show = VGSShow.Builder(reactApplicationContext, SharedConfig.VAULT_ID)
            .setEnvironment(SharedConfig.ENVIRONMENT.toVGSEnvironment())
            .build()
        show.addOnResponseListener(this)
    }

    override fun getName(): String = "VGSShowManager"

    override fun onResponse(response: VGSResponse) {
        sendResponse(response)
    }

    fun subscribe(view: VGSTextView) {
        show.subscribe(view)
    }

    @ReactMethod
    fun revealData(callback: Callback?) {
        this.callback = callback
        show.requestAsync("/post", VGSHttpMethod.POST, showPayload)
    }

    private fun sendResponse(response: VGSResponse) {
        val responseStr = StringBuilder("Code: ")
            .append(response.code)
        if (response is VGSResponse.Error) {
            responseStr
                .append("\n")
                .append("Error: ")
                .append(response.message)
        }
        callback!!.invoke(responseStr.toString())
    }
}
